### ISSUE TITLE
world-aware adding of particles to fluid system 

### ISSUE THREAD

**Issue by rezural** [rustsim/salva#10]
when adding particles to the system (either on initialization, or via testbed.add_callback_with_fluids), it is difficult know for certain that there is not a particle (or collider) already in this position. adding a particle that intersects causes explosions basically.

I am currently testing that none of the existant particles intersect new particles using ncollide3d::query::proximity_ball_ball, but this is sub optimal.

Is there a performant way of testing whether a particle is intersecting with the fluid at all?

**Comment by sebcrozet** [comment]
Hi! Testing if a particle exists at a specific location isn't implemented yet on Salva, so for now testing ball-ball proximity is the only option unfortunately.

Testing if there is a collider in a specific location is possible if you are using nphysics because you can use `geometrical_world.interferences_with_point`. And if you are using ncollide directly, you can do `collision_world.interferences_with_point`.

**Comment by rezural** [comment]
OK, Thanks.

Any ideas on data structures for efficient search through the particle space? Would this structure need to be updated for every particle movement? Would it be too prohibitive?

Cheers

**Comment by sebcrozet** [comment]
@rezural The ideal data structure would be a grid. Actually, this is what the `LiquidWorld` uses and maintains internally. So the ideally we would want to let the `LiquidWorld` tell us what particles lie in a given region of space.

Alternatively, you can maintain your own [HGrid](https://docs.rs/salva3d/0.4.0/salva3d/geometry/struct.HGrid.html) by clearing and re-adding points after each update and use [HGrid::cells_intersecting_aabb](https://docs.rs/salva3d/0.4.0/salva3d/geometry/struct.HGrid.html#method.cells_intersecting_aabb) to find all the particles in a given area (defined as the AABB of the particle you intend to add).

**Comment by nhemsley** [comment]
Right,

something like:

```
liquid_world::particles_within_shape(&self, Shape[Handle]) -> Vec<Point3<f32>
liquid_world::particles_within_shape_aabb(&self, Shape[Handle]) -> HGrid<..> // for efficient queries later on
```

?

**Comment by nhemsley** [comment]
I can create a PR for that if you'd like that in salva.

**Comment by rezural** [comment]
Right,

Something like?

```rust
liquid_world.particles_within(region: Shape, isometry) -> Vec<Point3<f32>>
liquid_world.particles_within_aabb(region: Shape, isometry) -> HGrid<..> // for efficient lookup 

```

I can probably add this in there of you'd like to see it.

**Comment by sebcrozet** [comment]
The following should be enough:

```rust
liquid_world.particles_intersecting_aabb(&self, aabb: &AABB<f32>) -> Vec<Point<f32>>;
```

In the end we would probably need to return some identifier that allows the user to retrieve the fluid containing said particle, but since fluid/boundary addressing will change soon, I think returning `Vec<Point<f32>>` is a good start and will cover your use-case.

> I can probably add this in there of you'd like to see it.

Yes, that would be great. Ideally, the PR should target the [rapier branch](https://github.com/dimforge/salva/tree/rapier).

**Comment by rezural** [comment]
So the rapier branch doesnt build against rapier (with rapier, ncollide, nalgebra cloned locally). 

You mentioned a salva branch on the rapier repository? I cant find that on the dimforge repo.

Otherwise I will make a pr against salva master, hopefully within a few days.

**Comment by sebcrozet** [comment]
> You mentioned a salva branch on the rapier repository? I cant find that on the dimforge repo.

That's right, I forgot to push it. Today I pushed it and merged it to the master branch of Rapier.

In addition I merged #11 which drops nphysics to use Rapier instead. So the `master` branch of salva now compiles and has integration with Rapier (using the github version of Rapier until it is released at the end of this month).

**Comment by rezural** [comment]
OK, thanks for that!

I will look into getting this working with the latest master.

**Comment by rezural** [comment]
I'm tempted to add a particles_intersecting_aabb method to the Fluid struct, and just collect these together in the liquid world, which will probably make this easier to work in with the addressing changes.. Thoughts?

**Comment by sebcrozet** [comment]
This will be implemented by #13
